### DISCUSSION TITLE
What is a decent strategy to refresh a `FutureProvider` when disposing another one?

### DISCUSSION THREAD

**Discussion by matuella** [rrousselGit/riverpod#1001] [Q&A]
Let's say that we have two providers: `providerA` and `providerB`. There are some scenarios where after disposing of `providerB`, I wanted to call `ref.refresh(providerA)`. If I'm not mistaken, there is only one way to know when a provider was disposed, and it's on the `create` function, `onDispose`. The problem is that when I call the `ref.refresh` on the dispose method, that particular `ref` is already disposed of, meaning it'll throw an error.

Thus, my question: how can I listen to some provider being disposed, from another reference?

**Comment by rrousselGit** [comment]
> There are some scenarios where after disposing of providerB, I wanted to call ref.refresh(providerA)

That sounds fishy. What exactly is this scenario? You most likely don't want to do that.

**Reply by matuella** [reply]
Let's assume that we are consuming a REST application that has a list of elements in one screen which navigates to each individual element. This second "screen" is using an autoDispose provider (let's call it `providerB`) which is only alive when this second screen is alive, so I can't add a listener on its parent (let's call it `providerA`), as it may or may not be alive. Now, when disposing of `providerB`, I wanted to refresh the `providerA` because changes on this second screen impacts on the previous one.

**Reply by matuella** [reply]
Knowing when a provider is (or not) alive is also something that would be valuable in this situation, but it's also not supported.

**Reply by rrousselGit** [reply]
> Now, when disposing of providerB, I wanted to refresh the providerA because changes on this second screen impacts on the previous one.

Why does this needs to be done "when providerB is disposed"? Isn't there a better place to do that? Such as when the second screen is poped

**Reply by matuella** [reply]
Sure, but we could argue that, let's say this second screen is popped in multiple possible ways, not necessarily in just a single "point", I would have to always go there and imperatively call the `providerB` to reload itself. This also opens problems for things like: "when a new "pop" method is added, I'd have to remember to also call all depending providers to refresh", which is a design flaw for me.

Having the ability to observe events like "disposals" and "being/not being alive" are useful for side-effects, but I'm unable to achieve a reactive approach with how things are structured.

I'm not saying that all things SHOULD be observable, it's just that, IMHO, it makes sense to be able to react to all of the provider lifecycle.

**Reply by matuella** [reply]
Sorry to bring up this discussion again, but what do you think about the current state of `riverpod` and the observability of a provider lifecycle?

**Reply by rrousselGit** [reply]
I don't get the question.

But for your request, that's not something allowed.

Providers are voluntarily restricted in various ways. This is a requirement for making them maintainable.

Refreshing a provider when another one is disposed is the wrong way to architecture things.
Rather, have the former watch the later.
